### PR TITLE
Add reward unlock function

### PR DESCRIPTION
**Pull request by eric-linear**
1. Added `unlockReward` function in `LnRewardLocker`
2. Added `collateralFromUnlockReward` function in `LnCollateralSystem`
3. Added corresponding test cases

### CODE REVIEW

File: contracts/LnCollateralSystem.sol [hunk 1]
```diff
@@ -14,6 +14,7 @@ import "./interfaces/ILnBuildBurnSystem.sol";
 import "./interfaces/ILnDebtSystem.sol";
 import "./interfaces/ILnConfig.sol";
 import "./interfaces/ILnRewardLocker.sol";
+import "./utilities/TransferHelper.sol";
```
Comment (xJonathanLEI) [inline contracts/LnCollateralSystem.sol:17]: Imports should be sorted.

Comment (eric-linear) [reply contracts/LnCollateralSystem.sol:17]: Updated in a583bf5

File: contracts/LnRewardLocker.sol [hunk 2]
```diff
@@ -124,6 +134,33 @@ contract LnRewardLocker is ILnRewardLocker, LnAdminUpgradeable {
         _moveRewardProRata(from, recipient1, amount1, recipient2, amount2, rewardEntryIds);
     }
 
+    function updateCollateralSystemAddress(address _collateralSystemAddr) external onlyAdmin {
+        require(_collateralSystemAddr != address(0), "Collateral system address must not be 0");
```
Comment (xJonathanLEI) [inline contracts/LnRewardLocker.sol:138]: Prefix all error messages with contract name.

File: contracts/LnRewardLocker.sol [hunk 3]
```diff
@@ -50,6 +52,9 @@ contract LnRewardLocker is ILnRewardLocker, LnAdminUpgradeable {
 
     bytes32 private constant ROLE_LOCK_REWARD = "LOCK_REWARD";
     bytes32 private constant ROLE_MOVE_REWARD = "MOVE_REWARD";
+    bytes32 private constant ROLE_UNLOCK_REWARD = "UNLOCK_REWARD";
+    address public collateralSystemAddr;
```
Comment (xJonathanLEI) [inline contracts/LnRewardLocker.sol:56]: Sort global variables with access specifier.

Constants are not state variables.

File: contracts/LnRewardLocker.sol [hunk 4]
```diff
@@ -61,6 +66,11 @@ contract LnRewardLocker is ILnRewardLocker, LnAdminUpgradeable {
         _;
     }
 
+    modifier onlyUnlockRewardRole() {
```
Comment (xJonathanLEI) [inline contracts/LnRewardLocker.sol:69]: The `unlockReward()` function should be permission-less. Why stop anyone from volunteering to help others unlock their rewards?

File: contracts/LnRewardLocker.sol (hunk 2 above)
Comment (eric-linear) [reply contracts/LnRewardLocker.sol]: Updated in 035a9a5

File: contracts/LnRewardLocker.sol (hunk 3 above)
Comment (eric-linear) [reply contracts/LnRewardLocker.sol]: Updated in 035a9a5

File: contracts/LnRewardLocker.sol (hunk 4 above)
Comment (eric-linear) [reply contracts/LnRewardLocker.sol]: Removed unlock reward role and updated in 035a9a5

File: tests/utilities/init.ts [hunk 5]
```diff
@@ -430,6 +437,16 @@ export const deployLinearStack = async (
     }
   );
 
+  /**
+   * Assign the following role to contract `LnRewardSystem`:
+   * - LOCK_REWARD
+   */
+  await lnAccessControl.connect(admin).SetRoles(
```
Comment (xJonathanLEI) [inline tests/utilities/init.ts:444]: This is no longer needed

Comment (eric-linear) [reply tests/utilities/init.ts:444]: I will need the LnRewardSystem contract to have the LOCK_REWARD role in order to add reward to LnRewardLocker in the unlock reward integration test.

Comment (xJonathanLEI) [reply tests/utilities/init.ts:444]: Yeah you're right. I thought this was `UNLOCK_REWARD`.

File: tests/LnCollateralSystem.spec.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+      .connect(admin)
+      .UpdateTokenInfos(
+        [ethers.utils.formatBytes32String("LINA")],
+        [linaToken.address],
+        [1],
+        [false]
+      );
+  });
+
+  it("only reward locker can call collateralFromUnlockReward function", async () => {
```
Comment (xJonathanLEI) [inline tests/LnCollateralSystem.spec.ts:103]: This case in its current form only tests that a non-reward-locker _cannot_ call this function. The "only" aspect is not fully tested in that it hasn't asserted that a legit reward-locker _can_ call it.

Please also check other test cases to see if similar issues exist.

**Comment by xJonathanLEI** [comment]
The implementation part looks good. Only small issues remain in tests. Please update accordingly. Thanks.

@eric-linear

File: tests/LnCollateralSystem.spec.ts (hunk 6 above)
Comment (eric-linear) [reply tests/LnCollateralSystem.spec.ts:103]: Oh right! Test cases are updated in af3a3f7.

File: tests/LnRewardLocker.spec.ts [hunk 7]
```diff
@@ -175,4 +200,182 @@ describe("LnRewardLocker", function () {
...
+      bob.address, // user
+      10, // amount
+      unlockTime.toSeconds() // unlockTime
+    );
+
+    expect(await lnRewardLocker.lockedAmountByAddresses(bob.address)).to.equal(
+      10
+    );
+
+    setNextBlockTimestamp(ethers.provider, unlockTime);
```
Comment (xJonathanLEI) [inline tests/LnRewardLocker.spec.ts:225]: `setNextBlockTimestamp()` returns `Promise`, which must be awaited.

There are multiple occurrences of this issue in this PR.

**Review by xJonathanLEI — APPROVED** [review]

**Pull request merged**